### ISSUE TITLE
NeuralSDE example throws "NotImplementedError: outfeed rewrite closed_call"

### ISSUE THREAD

**Issue by ciupakabra** [patrick-kidger/diffrax#176]
I'm not sure if this is a problem on diffrax or my setup side, but the NeuralSDE example throws a `NotImplementedError: outfeed rewrite closed_call`. A minimal working example is the following:

```python
import diffrax
import jax.numpy as jnp
import jax.random as jrandom
import jax.nn as jnn
import jax
    
def loss(alpha, dim, num_steps, key):
    dt0 = 1 / (num_steps + 1)
    ts = jnp.linspace(0, 1, num_steps + 1)

    control = diffrax.VirtualBrownianTree(
        t0=0, t1=1, tol=dt0 / 2, shape=(dim,), key=key,
    )
    
    drift_term = diffrax.ODETerm(lambda t, y, args: - alpha * y)
    diffusion_term = diffrax.ControlTerm(lambda t, y, args : alpha * jnp.ones((dim,)), control)

    terms = diffrax.MultiTerm(drift_term, diffusion_term)
    solver = diffrax.Euler()

    y0 = jnp.zeros((dim,))
    saveat = diffrax.SaveAt(ts=ts)

    sol = diffrax.diffeqsolve(
        terms, 
        solver, 
        0, 
        1, 
        dt0, 
        y0, 
        saveat=saveat, 
        max_steps=num_steps+1, 
    )
    return jnp.sum((sol.ys[-1]-1)**2)

@jax.grad
def grad_loss(alpha, dim, num_steps, key):
    return loss(alpha, dim, num_steps, key)

if __name__=="__main__":
    key = jrandom.PRNGKey(0)

    dim = 1
    num_steps = 20
    alpha = jnp.array((1.0,))

    loss(alpha, dim, num_steps, key)
    grad_loss(alpha, dim, num_steps, key)

```
which outputs:

```
Traceback (most recent call last):
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 48, in <module>
    grad_loss(alpha, dim, num_steps, key)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/api.py", line 1068, in grad_f
    _, g = value_and_grad_f(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/api.py", line 1144, in value_and_grad_f
    ans, vjp_py = _vjp(f_partial, *dyn_args, reduce_axes=reduce_axes)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/api.py", line 2579, in _vjp
    out_primal, out_vjp = ad.vjp(
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/interpreters/ad.py", line 133, in vjp
    out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/interpreters/ad.py", line 122, in linearize
    jaxpr, out_pvals, consts = pe.trace_to_jaxpr_nounits(jvpfun_flat, in_pvals)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/profiler.py", line 313, in wrapper
    return func(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 769, in trace_to_jaxpr_nounits
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/linear_util.py", line 168, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 38, in grad_loss
    return loss(alpha, dim, num_steps, key)
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 24, in loss
    sol = diffrax.diffeqsolve(
  File "/home/anonymous/env/lib/python3.10/site-packages/equinox/jit.py", line 96, in __call__
    return __self._fun_wrapper(False, args, kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/equinox/jit.py", line 92, in _fun_wrapper
    dynamic_out, static_out = self._cached(dynamic, static)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/api.py", line 604, in cache_miss
    out_flat = xla.xla_call(
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/core.py", line 1959, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/core.py", line 1975, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/core.py", line 701, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/dispatch.py", line 234, in _xla_call_impl
    compiled_fun = xla_callable(fun, device, backend, name, donated_invars,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/linear_util.py", line 306, in memoized_fun
    ans = call(fun, *args)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/dispatch.py", line 369, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars, False,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/profiler.py", line 313, in wrapper
    return func(*args, **kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/dispatch.py", line 468, in lower_xla_callable
    jaxpr = apply_outfeed_rewriter(jaxpr)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/_src/dispatch.py", line 588, in apply_outfeed_rewriter
    return outfeed_rewriter(jaxpr)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1806, in <lambda>
    dispatch.outfeed_rewriter = lambda j: _rewrite_jaxpr(j, False, False)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1569, in _rewrite_eqn
    new_jaxpr = _rewrite_closed_jaxpr(carry_jaxpr, True, True)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1466, in _rewrite_closed_jaxpr
    new_jaxpr = _rewrite_jaxpr(cjaxpr.jaxpr, has_input_token, has_output_token)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1556, in _rewrite_eqn
    branches=tuple(
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1557, in <genexpr>
    _rewrite_closed_jaxpr(jaxpr, True, True)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1466, in _rewrite_closed_jaxpr
    new_jaxpr = _rewrite_jaxpr(cjaxpr.jaxpr, has_input_token, has_output_token)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1689, in _rewrite_eqn
    raise NotImplementedError(f"outfeed rewrite {eqn.primitive}")
jax._src.traceback_util.UnfilteredStackTrace: NotImplementedError: outfeed rewrite closed_call

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 48, in <module>
    grad_loss(alpha, dim, num_steps, key)
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 38, in grad_loss
    return loss(alpha, dim, num_steps, key)
  File "/home/anonymous/tmp/diffrax-bug/example.py", line 24, in loss
    sol = diffrax.diffeqsolve(
  File "/home/anonymous/env/lib/python3.10/site-packages/equinox/jit.py", line 96, in __call__
    return __self._fun_wrapper(False, args, kwargs)
  File "/home/anonymous/env/lib/python3.10/site-packages/equinox/jit.py", line 92, in _fun_wrapper
    dynamic_out, static_out = self._cached(dynamic, static)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1806, in <lambda>
    dispatch.outfeed_rewriter = lambda j: _rewrite_jaxpr(j, False, False)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1569, in _rewrite_eqn
    new_jaxpr = _rewrite_closed_jaxpr(carry_jaxpr, True, True)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1466, in _rewrite_closed_jaxpr
    new_jaxpr = _rewrite_jaxpr(cjaxpr.jaxpr, has_input_token, has_output_token)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1556, in _rewrite_eqn
    branches=tuple(
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1557, in <genexpr>
    _rewrite_closed_jaxpr(jaxpr, True, True)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1466, in _rewrite_closed_jaxpr
    new_jaxpr = _rewrite_jaxpr(cjaxpr.jaxpr, has_input_token, has_output_token)
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1502, in _rewrite_jaxpr
    _rewrite_eqn(eqn, eqns, last_token_var, output_token_var,
  File "/home/anonymous/env/lib/python3.10/site-packages/jax/experimental/host_callback.py", line 1689, in _rewrite_eqn
    raise NotImplementedError(f"outfeed rewrite {eqn.primitive}")
NotImplementedError: outfeed rewrite closed_call

```

With `num_steps=10` the code completes successfully. Without the control term, i.e. a regular ODE, the code completes successfully as well. This was run on arch with GeForce GTX 1660 SUPER.

Relevant versions:

```
jax                          0.3.21
jaxlib                       0.3.20+cuda11.cudnn82
diffrax                      0.2.1
equinox                      0.8.0
```

**Comment by patrick-kidger** [comment]
Thanks for the report. I think this is due to JAX changing/breaking things upstream. I didn't realise this was affecting our examples, however.

I'll try and get a fix out for this in the next few days. As a temporary measure in the mean time you can probably switch to an earlier version of JAX.

**Comment by patrick-kidger** [comment]
This should now have been fixed in #190.
